File: apps/mac-gateway/src/protocol.py
from __future__ import annotations

from dataclasses import dataclass
from enum import IntEnum
import random
import time
from typing import Callable

MAX_PACKET_SIZE = 200
HEADER_SIZE = 4
MAX_PAYLOAD_SIZE = MAX_PACKET_SIZE - HEADER_SIZE
DEFAULT_REASSEMBLY_TIMEOUT_SECONDS = 60


class RouteId(IntEnum):
    HUMAN_CHAT = 0x01
    MCP_REQUEST = 0x02
    TELEGRAM = 0x03
    IOT_CONTROL = 0x04


@dataclass(frozen=True)
class DecodedPacket:
    route_id: int
    message_id: int
    chunk_index: int
    total_chunks: int
    payload_bytes: bytes
    payload_text: str


@dataclass(frozen=True)
class ReassembledMessage:
    route_id: int
    message_id: int
    payload_text: str
    total_chunks: int


@dataclass
class _PendingMessage:
    total_chunks: int
    chunks: dict[int, bytes]
    created_at: float


class ProtocolCodec:
    def __init__(self, *, rng: random.Random | None = None) -> None:
        self._rng = rng or random.Random()

    def compress(self, payload_text: str) -> bytes:
        return payload_text.encode("utf-8")

    def decompress(self, payload_bytes: bytes) -> str:
        return payload_bytes.decode("utf-8")

    def encode_chunks(self, route_id: int, payload_text: str, message_id: int | None = None) -> list[bytes]:
        if not (0 <= route_id <= 0xFF):
            raise ValueError("route_id must fit in one byte")

        compressed = self.compress(payload_text)
        if not compressed:
            data_chunks = [b""]
        else:
            data_chunks = [
                compressed[i : i + MAX_PAYLOAD_SIZE]
                for i in range(0, len(compressed), MAX_PAYLOAD_SIZE)
            ]

        msg_id = self._rng.randint(0, 255) if message_id is None else message_id
        if not (0 <= msg_id <= 0xFF):
            raise ValueError("message_id must fit in one byte")

        total = len(data_chunks)
        packets: list[bytes] = []
        for idx, chunk in enumerate(data_chunks, start=1):
            header = bytes((route_id, msg_id, idx, total))
            packets.append(header + chunk)
        return packets

    def decode_packet(self, packet: bytes) -> DecodedPacket:
        if len(packet) < HEADER_SIZE:
            raise ValueError("Packet must be at least 4 bytes")

        route_id = packet[0]
        message_id = packet[1]
        chunk_index = packet[2]
        total_chunks = packet[3]

        if chunk_index < 1:
            raise ValueError("chunk_index must be >= 1")
        if total_chunks < 1:
            raise ValueError("total_chunks must be >= 1")
        if chunk_index > total_chunks:
            raise ValueError("chunk_index cannot exceed total_chunks")

        payload_bytes = packet[HEADER_SIZE:]
        payload_text = self.decompress(payload_bytes)
        return DecodedPacket(
            route_id=route_id,
            message_id=message_id,
            chunk_index=chunk_index,
            total_chunks=total_chunks,
            payload_bytes=payload_bytes,
            payload_text=payload_text,
        )


class ReassemblyBuffer:
    def __init__(
        self,
        *,
        codec: ProtocolCodec,
        timeout_seconds: int = DEFAULT_REASSEMBLY_TIMEOUT_SECONDS,
        now_fn: Callable[[], float] = time.monotonic,
    ) -> None:
        self._codec = codec
        self._timeout_seconds = timeout_seconds
        self._now_fn = now_fn
        self._pending: dict[tuple[int, int], _PendingMessage] = {}

    def add_packet(self, packet: bytes) -> ReassembledMessage | None:
        decoded = self._codec.decode_packet(packet)
        key = (decoded.route_id, decoded.message_id)
        now = self._now_fn()

        existing = self._pending.get(key)
        if existing is None:
            existing = _PendingMessage(total_chunks=decoded.total_chunks, chunks={}, created_at=now)
            self._pending[key] = existing
        elif existing.total_chunks != decoded.total_chunks:
            raise ValueError("Received packet with inconsistent total chunks")

        existing.chunks[decoded.chunk_index] = decoded.payload_bytes

        if len(existing.chunks) != existing.total_chunks:
            return None

        joined = b"".join(existing.chunks[idx] for idx in range(1, existing.total_chunks + 1))
        payload_text = self._codec.decompress(joined)
        del self._pending[key]

        return ReassembledMessage(
            route_id=decoded.route_id,
            message_id=decoded.message_id,
            payload_text=payload_text,
            total_chunks=existing.total_chunks,
        )

    def expire_stale(self) -> int:
        now = self._now_fn()
        stale_keys = [
            key
            for key, pending in self._pending.items()
            if (now - pending.created_at) > self._timeout_seconds
        ]
        for key in stale_keys:
            del self._pending[key]
        return len(stale_keys)

File: apps/web-client/src/lib/protocol.ts
export const MAX_PACKET_SIZE = 200
export const HEADER_SIZE = 4
export const MAX_PAYLOAD_SIZE = MAX_PACKET_SIZE - HEADER_SIZE

export enum RouteId {
  HumanChat = 0x01,
  McpRequest = 0x02,
  Telegram = 0x03,
  IotControl = 0x04,
}

export type DecodedPacket = {
  routeId: number
  messageId: number
  chunkIndex: number
  totalChunks: number
  payloadBytes: Uint8Array
  payloadText: string
}

export type ReassembledMessage = {
  routeId: number
  messageId: number
  payloadText: string
  totalChunks: number
}

type PendingMessage = {
  totalChunks: number
  chunks: Map<number, Uint8Array>
  createdAt: number
}

export class ProtocolCodec {
  private readonly enc = new TextEncoder()
  private readonly dec = new TextDecoder()

  compress(text: string): Uint8Array {
    return this.enc.encode(text)
  }

  decompress(data: Uint8Array): string {
    return this.dec.decode(data)
  }

  encodeChunks(routeId: number, payloadText: string, messageId: number): Uint8Array[] {
    if (routeId < 0 || routeId > 255) {
      throw new Error('routeId must fit in one byte')
    }
    if (messageId < 0 || messageId > 255) {
      throw new Error('messageId must fit in one byte')
    }

    const compressed = this.compress(payloadText)
    const dataChunks = compressed.length
      ? Array.from({ length: Math.ceil(compressed.length / MAX_PAYLOAD_SIZE) }, (_, i) =>
          compressed.slice(i * MAX_PAYLOAD_SIZE, (i + 1) * MAX_PAYLOAD_SIZE),
        )
      : [new Uint8Array()]

    return dataChunks.map((chunk, idx) => {
      const header = new Uint8Array([routeId, messageId, idx + 1, dataChunks.length])
      const packet = new Uint8Array(header.length + chunk.length)
      packet.set(header)
      packet.set(chunk, HEADER_SIZE)
      return packet
    })
  }

  decodePacket(packet: Uint8Array): DecodedPacket {
    if (packet.length < HEADER_SIZE) {
      throw new Error('Packet must be at least 4 bytes')
    }

    const routeId = packet[0]
    const messageId = packet[1]
    const chunkIndex = packet[2]
    const totalChunks = packet[3]

    if (chunkIndex < 1) {
      throw new Error('chunkIndex must be >= 1')
    }
    if (totalChunks < 1) {
      throw new Error('totalChunks must be >= 1')
    }
    if (chunkIndex > totalChunks) {
      throw new Error('chunkIndex cannot exceed totalChunks')
    }

    const payloadBytes = packet.slice(HEADER_SIZE)
    return {
      routeId,
      messageId,
      chunkIndex,
      totalChunks,
      payloadBytes,
      payloadText: this.decompress(payloadBytes),
    }
  }
}

export class ReassemblyBuffer {
  private readonly pending = new Map<string, PendingMessage>()

  constructor(
    private readonly codec: ProtocolCodec,
    private readonly timeoutSeconds = 60,
    private readonly nowFn: () => number = () => Date.now() / 1000,
  ) {}

  addPacket(packet: Uint8Array): ReassembledMessage | null {
    const decoded = this.codec.decodePacket(packet)
    const key = `${decoded.routeId}-${decoded.messageId}`
    const existing = this.pending.get(key)

    if (!existing) {
      this.pending.set(key, {
        totalChunks: decoded.totalChunks,
        chunks: new Map([[decoded.chunkIndex, decoded.payloadBytes]]),
        createdAt: this.nowFn(),
      })
      return decoded.totalChunks === 1
        ? this.completeAndRemove(key, decoded.routeId, decoded.messageId)
        : null
    }

    if (existing.totalChunks !== decoded.totalChunks) {
      throw new Error('Received packet with inconsistent total chunks')
    }

    existing.chunks.set(decoded.chunkIndex, decoded.payloadBytes)
    if (existing.chunks.size !== existing.totalChunks) {
      return null
    }

    return this.completeAndRemove(key, decoded.routeId, decoded.messageId)
  }

  expireStale(): number {
    const now = this.nowFn()
    const stale = [...this.pending.entries()]
      .filter(([, pending]) => now - pending.createdAt > this.timeoutSeconds)
      .map(([key]) => key)

    stale.forEach((key) => this.pending.delete(key))
    return stale.length
  }

  private completeAndRemove(key: string, routeId: number, messageId: number): ReassembledMessage {
    const pending = this.pending.get(key)!

    const payload = new Uint8Array(
      [...Array.from({ length: pending.totalChunks }, (_, i) => i + 1)]
        .map((idx) => pending.chunks.get(idx)!)
        .reduce((acc, cur) => acc + cur.length, 0),
    )

    let offset = 0
    for (let i = 1; i <= pending.totalChunks; i += 1) {
      const chunk = pending.chunks.get(i)!
      payload.set(chunk, offset)
      offset += chunk.length
    }

    this.pending.delete(key)
    return {
      routeId,
      messageId,
      totalChunks: pending.totalChunks,
      payloadText: this.codec.decompress(payload),
    }
  }
}
